docs/modules/ROOT/pages/upgrades.adoc

```diff
@@ -0,0 +1,113 @@
+:contract_class: https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/contract-classes/[Contract Class]
+:class_hash: https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/class-hash/[class hash]
+:replace_class_syscall: https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/system-calls-cairo1/#replace_class[replace_class]
+:upgradeable: https://github.com/OpenZeppelin/cairo-contracts/blob/cairo-2/src/upgrades/upgradeable.cairo[Upgradeable]
+:ownable: xref:access.adoc#ownership_and_ownable[Ownable]
+
+= Upgradeability
+
+Smart contracts are immutable by design, preventing any update after the contract is deployed.
+This supports decentralization and trustlessness, but in certain cases can be an drawback for example
+by removing the ability of fixing bugs or vulnerabilities.
+
+Multiple patterns have been developed for making a contract upgradeable including the widely adopted proxy pattern.
+
+Starknet simplifies upgradeability by offering a syscall that "updates" the contract source code, removing the need for proxies.
+
+== Replacing Contract Classes
+
+It's important to understand the difference between a contract declaration and a contract instance in order to better comprehend how upgradeability works in Cairo.
+A contract must first be declared before it can be deployed.
+After deployment, the contract address is then associated to the previously
+declared {contract_class}, by passing the corresponding {class_hash}, becoming an "instance" of this "class".
+The Starknet stack will use this class for processing transactions going through the contract.
+
+=== `replace_class_syscall`
+
+As mentioned, this class hash determines the contract source code, which is designed to be immutable.
+However, since sometimes it is necessary to update the contract source code, the `{replace_class_syscall}` syscall
+was introduced. This syscall allows developers to replace the class hash of an already deployed contract.
```

Comment:
If they don't implement `replace_class_hash` and library call (or if the library call is set to an implementation that doesn't use `replace_class_hash`), they are. Immutability is one of the main characteristics for making blockchain suited for trustlessness applications. As mentioned in the other comment, the name smart contract comes in part from the fact that they are immutable by design (even when you can overcome this), mentioning immutability is important imo.

Reply:
I'd extend it to something like: "designed to be immutable by default" which better implies that devs must opt in for mutable contracts

Reply:
continued the conversation in another thread, but i'll address a slightly different topic on this one: that source code immutability is not a core property of blockchains.

the main and biggest example is that we effectively have upgrades in the main and in most smart contract blockchains out there, and yet still manages to build complete legitimacy over its code and execution-- being legitimacy the core pillar and goal of building a blockchain. so source code immutability is not really a requirement, but clear rules and verifiable execution are. upgrades through proxies or source code replacement can only happen through previously encoded rules, predictable, with known conditions, and verifiable execution.

the bottomline of this discussion to me is not about philosophical/theoretical definitions of what smart contracts or blockchains should be, but to describe the systems we're building on and smart contracts in Starknet are undeniably mutable. They can mutate themselves by running the `replace_class` syscall, or they can mutate if such syscall happens during a `library_call`. If to avoid upgrades you need to carefully analyze a contract's code and execution path to make sure none of these syscalls is being executed that means they're not "immutable by default".

as a side note, i wouldn't say blockchains are trustless. they're trust minimization technologies and they are good to reason about where trust is (in this or that actor, in this or that piece of hardware/software), but they don't fully remove it.

Reply:
Your call on this, but a couple of last comments form my side:

Smart Contracts in Ethereum are immutable by design, you cannot change the code of a smart contract, not even by using proxies. Proxy is an application-level mechanism for changing the storage with a different logic, but the code of the deployed smart contract remains the same. In Starknet is true they broke this by allowing replacing classes, but I wouldn't say they're "immutable if you take these precautions", I would say they are immutable unless you add this mechanism. 

In Ethereum they use the word immutable and trustlessness a lot, and I think it makes sense, even when we can get really into the technicalities but that's not the point of this comment.

You can check the first two paragraphs here as an example:
https://ethereum.org/en/developers/docs/smart-contracts/upgrading/

I will remove the words after you confirm, but I still feel they are not wrong, and they add value.

Reply:
blockchain immutability is not the same as smart contract code immutability. the former is in fact an integral component of the value proposition because execution and state finality is indeed required to build legitimacy, but the latter is not. even if the proxy code stays immutable, what's the value of such immutability if the executed code in runtime is in fact mutable?

> I would say they are immutable unless you add this mechanism

i don't fully understand this framing, i might be missing something. how if you have a button to do "X", why saying the protocol is "no-X"? it's like saying smart contracts have immutable and non-modifiable storage unless you use `my_storage.write(...)`. it makes no sense to me.

Reply:
blockchain immutability is not the same as smart contract code immutability. the former is in fact an integral component of the value proposition because execution and state finality is indeed required to build legitimacy, but the latter is not. even if the proxy code stays immutable, what's the value of such immutability if the executed code in runtime is in fact mutable?

> I would say they are immutable unless you add this mechanism

i don't fully understand this framing, i might be missing something. how if you have a button to do "X", why saying the protocol is "no-X"? it's like saying smart contracts have immutable and non-modifiable storage unless you use `my_storage.write(...)`. it makes no sense to me.

> You can check the first two paragraphs here as an example:

sure I do recognize that in Ethereum there was a strong movement to make them immutable and they indeed made them "immutable by design", but that was an Ethereum design choice, not a core component of the "smart contract" definition, it doesn't mean that they as a general category have to be immutable. that's not even widely accepted in the Ethereum ecosystem and people hacked their way to proxy-based upgrades.

I'm open to understand the value you think this paragraph adds, i'm just failing to see it yet.